File: buffer/gpu/src/jvmMain/java/com/wsr/gpu/elementwise/operation/JMinus.java
package com.wsr.gpu.elementwise.operation;

public class JMinus {
    public native void minusD0ToD1(float x, long y, long result, long runtime);

    public native void minusD1ToD0(long x, float y, long result, long runtime);

    public native void minusD1ToD1(long x, long y, long result, long runtime);

    public native void minusD1ToD2(long x, long y, int yi, int yj, int axis, long result, long runtime);

    public native void minusD1ToD3(long x, long y, int yi, int yj, int yk, int axis, long result, long runtime);

    public native void minusD2ToD1(long x, int xi, int xj, long y, int axis, long result, long runtime);

    public native void minusD2ToD3(
            long x,
            int xi,
            int xj,
            long y,
            int yi,
            int yj,
            int yk,
            int axis1,
            int axis2,
            long result,
            long runtime
    );

    public native void minusD3ToD1(long x, int xi, int xj, int xk, long y, int axis, long result, long runtime);

    public native void minusD3ToD2(
            long x,
            int xi,
            int xj,
            int xk,
            long y,
            int yi,
            int yj,
            int axis1,
            int axis2,
            long result,
            long runtime
    );

    public native void minusD3ToD4(
            long x,
            int xi,
            int xj,
            int xk,
            long y,
            int yi,
            int yj,
            int yk,
            int yl,
            int axis1,
            int axis2,
            int axis3,
            long result,
            long runtime
    );

    public native void minusD4ToD1(long x, int xi, int xj, int xk, int xl, long y, int axis, long result, long runtime);

    public native void minusD4ToD2(
            long x,
            int xi,
            int xj,
            int xk,
            int xl,
            long y,
            int yi,
            int yj,
            int axis1,
            int axis2,
            long result,
            long runtime
    );

    public native void minusD4ToD3(
            long x,
            int xi,
            int xj,
            int xk,
            int xl,
            long y,
            int yi,
            int yj,
            int yk,
            int axis1,
            int axis2,
            int axis3,
            long result,
            long runtime
    );
}

File: buffer/cpu/src/jvmAndroidMain/kotlin/com/wsr/knist/cpu/elementwise/operation/minus/JMinus.kt
package com.wsr.knist.cpu.elementwise.operation.minus


object JMinus {
    external fun minusD0ToD1(x: Float, y: Long, result: Long)

    external fun minusD1ToD0(x: Long, y: Float, result: Long)
    external fun minusD1ToD1(x: Long, y: Long, result: Long)
    external fun minusD1ToD2(x: Long, y: Long, yi: Int, yj: Int, axis: Int, result: Long)
    external fun minusD1ToD3(x: Long, y: Long, yi: Int, yj: Int, yk: Int, axis: Int, result: Long)

    external fun minusD2ToD1(x: Long, xi: Int, xj: Int, y: Long, axis: Int, result: Long)
    external fun minusD2ToD3(
        x: Long,
        xi: Int,
        xj: Int,
        y: Long,
        yi: Int,
        yj: Int,
        yk: Int,
        axis1: Int,
        axis2: Int,
        result: Long,
    )

    external fun minusD3ToD1(x: Long, xi: Int, xj: Int, xk: Int, y: Long, axis: Int, result: Long)
    external fun minusD3ToD2(
        x: Long,
        xi: Int,
        xj: Int,
        xk: Int,
        y: Long,
        yi: Int,
        yj: Int,
        axis1: Int,
        axis2: Int,
        result: Long,
    )
    external fun minusD3ToD4(
        x: Long,
        xi: Int,
        xj: Int,
        xk: Int,
        y: Long,
        yi: Int,
        yj: Int,
        yk: Int,
        yl: Int,
        axis1: Int,
        axis2: Int,
        axis3: Int,
        result: Long,
    )

    external fun minusD4ToD1(
        x: Long,
        xi: Int,
        xj: Int,
        xk: Int,
        xl: Int,
        y: Long,
        axis: Int,
        result: Long,
    )
    external fun minusD4ToD2(
        x: Long,
        xi: Int,
        xj: Int,
        xk: Int,
        xl: Int,
        y: Long,
        yi: Int,
        yj: Int,
        axis1: Int,
        axis2: Int,
        result: Long,
    )
    external fun minusD4ToD3(
        x: Long,
        xi: Int,
        xj: Int,
        xk: Int,
        xl: Int,
        y: Long,
        yi: Int,
        yj: Int,
        yk: Int,
        axis1: Int,
        axis2: Int,
        axis3: Int,
        result: Long,
    )
}
